src/XMakeBuildEngine/UnitTestsPublicOM/Definition/ProjectItem_Tests.cs

```diff
@@ -663,26 +663,138 @@ private static void TestIncludeExclude(string projectContents, string[] inputFil
+        [InlineData(ItemWithIncludeAndExclude,
+            "a",
+            "dir",
+
+            "a",
+            "",
+
+            new[]
+            {
+                @"project\dir\a",
+            },
+            true
+            )]
+        // exclude does not match include; file is above the project file; exclude points next to the project file
+        [InlineData(ItemWithIncludeAndExclude,
+            "a",
+            "..",
+
+            "a",
+            "",
+
+            new[]
+            {
+                "a",
+            },
+            true
+            )]
+        public void IncludeAndExcludeWorkWithRelativeAndAbsolutePaths(
+            string projectContents,
+            string includeItem,
+            string includeRelativePath,
+            string excludeItem,
+            string excludeRelativePath,
+            string[] inputFiles,
+            bool includeSurvivesExclude)
+        {
+            Func<bool, string, string, string, string> adjustFilePath = (isAbsolute, testRoot, relativeFragmentFromRootToFile, file) =>
+                isAbsolute
+                    ? Path.GetFullPath(Path.Combine(testRoot, relativeFragmentFromRootToFile, file))
+                    : Path.Combine(relativeFragmentFromRootToFile, file);
+
+            Action<bool, bool> runTest = (includeIsAbsolute, excludeIsAbsolute) =>
+            {
+                var testRoot = "";
+                try
+                {
+                    string[] createdFiles;
+                    string projectFile;
+
+                    // projectContents is incomplete because we do not yet know the test directory paths
+                    testRoot = Helpers.CreateProjectInTempDirectoryWithFiles(projectContents, inputFiles, out projectFile, out createdFiles, "project");
+
+                    var projectFileDir = Path.GetDirectoryName(projectFile);
+
+                    var include = adjustFilePath(includeIsAbsolute, projectFileDir, includeRelativePath, includeItem);
+                    var exclude = adjustFilePath(excludeIsAbsolute, projectFileDir, excludeRelativePath, excludeItem);
+
+                    // includes and exclude may be absolute, so we can only format the project after we have the test directory paths
+                    var formattedProject = string.Format(projectContents, include, exclude);
+                    File.WriteAllText(projectFile, formattedProject);
+
+                    var expectedInclude = includeSurvivesExclude ? new[] {include} : new string[0];
+
+                    ObjectModelHelpers.AssertItems(expectedInclude, new Project(projectFile).Items.ToList());
+                }
+                finally
+                {
+                    ObjectModelHelpers.DeleteDirectory(testRoot);
+                    Directory.Delete(testRoot);
+                }
+            };
+
+            runTest(true, false);
+            runTest(false, true);
+            runTest(true, true);
```

Comment:
If you adjust one of these to fail, is the error understandable? It might be more diagnosable to split each combination here into different test cases.